core/templates/dev/head/player/InteractionService.js

```diff
@@ -0,0 +1,94 @@
+
+oppia.factory('interactionService', [
+    'warningsData', 'DEFAULT_ANSWER_GROUP_CLASSIFICATION_THRESHOLD', function(
+    warningsData, DEFAULT_ANSWER_GROUP_CLASSIFICATION_THRESHOLD) {
+  var stringifyClassifiedRule = function(rule) {
+    var paramList = [];
+    for (var key in rule.inputs) {
+      paramList.push(rule.inputs[key]);
+    }
+    var stringifiedRule = rule.rule_type;
+    stringifiedRule += '(' + paramList.join(',') + ')';
+    return stringifiedRule;
+  };
+  return {
+    'classifyAnswer': function(
+      answer, answerClassificationData, RULES) {
+      var answerGroups = answerClassificationData.answerGroups;
+      var defaultOutcome = answerClassificationData.defaultOutcome;
+
+      // Find the first group that satisfactorily matches the given answer. This
+      // is done by ORing (maximizing) all truth values of all rules over all
+      // answer groups. The group with the highest truth value is considered the
+      // best match.
+      var bestMatchedAnswerGroup;
+      var bestMatchedAnswerGroupIndex = answerGroups.length;
+      var bestMatchedRuleSpec;
+      var bestMatchedTruthValue = 0.0;
+      for (var i = 0; i < answerGroups.length; i++) {
+        var oredTruthValue = 0.0;
+        var bestRuleSpec;
+        for (var j = 0; j < answerGroups[i].rule_specs.length; j++) {
+          var ruleSpec = answerGroups[i].rule_specs[j];
+          var evaluatedTruthValue = RULES[ruleSpec.rule_type](
+            answer, ruleSpec.inputs);
+          if (evaluatedTruthValue > oredTruthValue) {
+            oredTruthValue = evaluatedTruthValue;
+            bestRuleSpec = ruleSpec;
+          }
+        }
+        if (oredTruthValue > bestMatchedTruthValue) {
+          bestMatchedTruthValue = oredTruthValue;
+          bestMatchedRuleSpec = bestRuleSpec;
+          bestMatchedAnswerGroup = answerGroups[i];
+          bestMatchedAnswerGroupIndex = i;
+        }
+      }
+
+      // The best matched group must match above a certain threshold. If no
+      // group meets this requirement, then the default 'group' automatically
+      // matches resulting in the outcome of the answer being the default
+      // outcome of the state.
+      var DEFAULT_RULESPEC_STRING = 'Default';
+      if (bestMatchedTruthValue >=
+          DEFAULT_ANSWER_GROUP_CLASSIFICATION_THRESHOLD) {
+        return {
+          'outcome': bestMatchedAnswerGroup.outcome,
+          'rule_spec_string': stringifyClassifiedRule(bestMatchedRuleSpec),
+          'answer_group_index': bestMatchedAnswerGroupIndex,
+          'classification_certainty': bestMatchedTruthValue
+        };
+      } else if (defaultOutcome !== undefined) {
+        return {
+          'outcome': defaultOutcome,
+          'rule_spec_string': DEFAULT_RULESPEC_STRING,
+          'answer_group_index': answerGroups.length,
+          'classification_certainty': 0.0
+        };
+      } else {
+        warningsData.addWarning('No default rule found.');
```

Comment:
This scenario is a safety net in case something seriously goes wrong somewhere else in the codebase. I recommend throwing an Error instead of just adding a warning. This should also be tested to ensure a scenario where a totally broken exploration cannot cause undefined behavior from this service (e.g. an exploration with a missing default outcome).